### DISCUSSION TITLE
Multiblock close plot

### DISCUSSION THREAD

**Discussion by chcarra** [pyvista/pyvista#4224] [Q&A]
For a particular scalar of a multiblock, I generate multiple plots offscreen and take screenshots. However, it gets to a point where I get the warning of the inability for the plot to render. I am assuming I need to close all the plots that I generated, but when I specify multiblock.close() it does not have that method. How do I close the figures that I have plotted in the background?

**Comment by banesullivan** [comment]
Would you please share some code and share the exact warning you are getting?

**Reply by chcarra** [reply]
```
block = exodus2pyvista('results_output.e', nstep=1)

grad_strn_energy = block.compute_derivative(scalars="nodal_strain_energy", preference='point')["gradient"]

block["grad_strain_energy_x"] = -grad_strn_energy[:, 0]
block["grad_strain_energy_y"] = -grad_strn_energy[:, 1]
block["grad_strain_energy_z"] = -grad_strn_energy[:, 2]

block["triaxiality_2"] = block["nodal_hydrostatic_stress"] / block["nodal_von_mises"]

newblock = block.slice(normal=(0, 1, 0))

pv.scalar_bars

pv.OFF_SCREEN = True

newblock.plot(scalars='grad_strain_energy_x', cmap='jet', cpos=(0, -1, 0), screenshot='grad_strain_energy_x_010.png')
```

I run this chunk of code in a particular loop and at some point I get the warning: "A screenshot is unable to be taken as the render window is not current or rendering is suppressed." My guess is that after running newblock.plot() so many times, I cannot render in the background

**Reply by banesullivan** [reply]
Hm. Thank you for clarifying the exact error message. Would you also share a "Report"? Further, are you running in Jupyter? If so, do you know which Jupyter backend is enabled?

```py
import pyvista as pv
print(pv.Report())
```

I suspect this is in Jupyter with the trame or ipyvtklink backends which if creating many plots will eventually eat up all of your resources. If this happens, your best bet is to use the `close_all()` API:

```py
import pyvista as pv
pv.close_all()
```

This may or may not help as I think the jupyter backends will still keep a reference to many of the plots

**Reply by chcarra** [reply]
Hello, thanks for the response. I am not running this in Jupyter. Attached is the report. I attempted pv.close_all(), but based on the eventual slowing down and eventually ceasing of producing figure, I assumed the offscreen plots are still there.

--------------------------------------------------------------------------------
  Date: Tue Apr 04 14:19:02 2023 PDT

                OS : Darwin
            CPU(s) : 12
           Machine : x86_64
      Architecture : 64bit
       Environment : Python
        GPU Vendor : ATI Technologies Inc.
      GPU Renderer : AMD Radeon Pro 560X OpenGL Engine
       GPU Version : 4.1 ATI-4.10.12

  Python 3.11.2 (main, Mar 24 2023, 00:31:37) [Clang 14.0.0
  (clang-1400.0.29.202)]

           pyvista : 0.38.5
               vtk : 9.2.6
             numpy : 1.24.2
           imageio : 2.27.0
            scooby : 0.7.1
             pooch : v1.7.0
        matplotlib : 3.7.1
             PyQt5 : 5.15.7
           IPython : 8.11.0
             scipy : 1.10.1
--------------------------------------------------------------------------------

**Reply by chcarra** [reply]
Another thing I tried. Was to run the following example code line by line.
`import numpy as np
import pyvista

point_cloud = np.random.random((100, 3))
pdata = pyvista.PolyData(point_cloud)
pdata['orig_sphere'] = np.arange(100)

sphere = pyvista.Sphere(radius=0.02, phi_resolution=10, theta_resolution=10)
pc = pdata.glyph(scale=False, geom=sphere, orient=False)
pc.plot(cmap='Reds')
pyvista.close_all()`

I noticed that the window did not close until after I quit python

**Reply by chcarra** [reply]
It looks like this has been issue that has occurred before. I saw something about implementing an older version of vtk, but I am not sure if that's still the suggestion.

**Reply by gattia** [reply]
Any updates here? I get a similar error. Im loading a mesh, creating a plotter, taking screenshots from 4 perspectives, closing the plotter (after deep_clean) and then going on to the next one. I have ~9000 meshes I want to analyze and it currently fails at ~ 160). See below for relevant code. 

I've also tried updating the below code to catch the error I get (`pyvista.plotting.errors.RenderWindowUnavailable: Render window is not current.`) doing `pyvista.close_all()`, waiting a few seconds, and then continuing and it doesn't fix it - I then just get the same error for ever loop. 

Note - this is in a .py file (not jupyter notebook). 


```python
for idx, mesh_path in enumerate(list_new_mesh_paths):
    print(f'{idx+1}/{n_meshes}')
    tic = time.time()
    filename = os.path.basename(mesh_path)
    subject = filename.split('_')[0]
    leg = filename.split('_')[1]
    
    if os.path.exists(os.path.join(folder_save, f'{subject}_{leg}_zx.png')):
        continue

    plotter = pv.Plotter(off_screen=True, notebook=False)
    mesh = pv.read(mesh_path)
    plotter.add_mesh(mesh, color="orange", lighting=True)
    for position in ['xy', 'xz', 'yx', 'zx']:  #['xz', 'zx', 'yz', 'zy', 'xy', 'yx', 'zx']
        plotter.camera_position = position
        plotter.screenshot(os.path.join(folder_save, f'{subject}_{leg}_{position}.png'))

    plotter.deep_clean()
    plotter.close()
    toc = time.time()
    print(f'Elapsed time: {toc-tic:.2f} seconds')

```

**Reply by gattia** [reply]
Alright... can confirm a fix. Don't remake the plotter in the loop and remove it out. However, there are some caveats. I first tried using `plotter.clear()` at the end of each loop, but that updates things so that there is no more lighting. A simple switch to `plotter.clear_actor()` seems to have fixed that problem.... going on 500 meshes in a row now (was crashing at 160 before).
